### PR TITLE
Add email link sign in support

### PR DESCRIPTION
**Pull request by renkelvin**

### CODE REVIEW

File: FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 1]
```diff
@@ -350,7 +537,7 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
             presentingViewController:(UIViewController *)presentingViewController
                         signInResult:(_Nullable FIRAuthResultCallback)result {
   id<FUIAuthDelegate> delegate = self.authUI.delegate;
-  [self.authUI.auth fetchProvidersForEmail:email
+  [self.authUI.auth fetchSignInMethodsForEmail:email
                          completion:^(NSArray<NSString *> *_Nullable providers,
```
Comment (Yue-Wang-Google) [inline FirebaseEmailAuthUI/FUIEmailAuth.m]: align colons

File: FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 2]
```diff
@@ -399,6 +586,39 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
...
+               actionHandler:^{
+                 [self generateURLParametersAndLocalCache:email
+                                          linkingProvider:newCredential.provider];
+                 [self.authUI.auth sendSignInLinkToEmail:email
+                                      actionCodeSettings:self.actionCodeSettings
+                                              completion:^(NSError * _Nullable error) {
+                                                [FUIAuthBaseViewController
+                                                    showAlertWithTitle:@"Sign-in email sent"
+                                                               message:[NSString stringWithFormat:@"A sign-in email with additional instrucitons was sent to %@. Check your email to complete sign-in.", email]
+                                              presentingViewController:nil];
```
Comment (Yue-Wang-Google) [inline FirebaseEmailAuthUI/FUIEmailAuth.m]: exceed 100 chars. make a local vairable.

File: FirebaseEmailAuthUI/FUIEmailAuth_Internal.h [hunk 3]
```diff
@@ -42,6 +42,12 @@ NS_ASSUME_NONNULL_BEGIN
 + (UIAlertController *)alertControllerForError:(NSError *)error
                                  actionHandler:(nullable FUIAuthAlertActionHandler)actionHandler;
 
+/** @fn generateURLParametersAndLocalCache:linkingProvider:
+    @brief Generate the parameters before sending out the email link. Append the parameters to
+        continue url and store them locally.
```
Comment (Yue-Wang-Google) [inline FirebaseEmailAuthUI/FUIEmailAuth_Internal.h]: indent

also add params

File: FirebaseEmailAuthUI/FUIEmailEntryViewController.m [hunk 4]
```diff
@@ -196,17 +211,63 @@ - (void)onNext:(NSString *)emailText {
...
+                actionCodeSettings:emailAuth.actionCodeSettings
+                        completion:^(NSError * _Nullable error) {
+    [self decrementActivity];
+
+    if (error) {
+      [FUIAuthBaseViewController showAlertWithTitle:@"Error"
+                                            message:error.description
+                           presentingViewController:self];
+    } else {
+      NSString *successMessage = [NSString stringWithFormat: @"A sign-in email with additional instrucitons was sent to %@. Check your email to complete sign-in.", email];
```
Comment (Yue-Wang-Google) [inline FirebaseEmailAuthUI/FUIEmailEntryViewController.m]: 100 chars.

File: FirebaseEmailAuthUI/FUIConfirmEmailViewController.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+/** @fn onNext:
+    @brief Should be called when user entered email. Triggers email verification before
+    pushing new controller
+    @param emailText Email value entered by user.
+ */
+- (void)onNext:(NSString *)emailText;
+
+/** @fn didChangeEmail:
+    @brief Should be called after any change of email value. Updates UI controls state
+    (e g state of next button)
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply FirebaseEmailAuthUI/FUIConfirmEmailViewController.h]: Remove the comments?

File: FirebaseEmailAuthUI/FUIConfirmEmailViewController.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+/** @fn onNext:
+    @brief Should be called when user entered email. Triggers email verification before
+    pushing new controller
+    @param emailText Email value entered by user.
+ */
+- (void)onNext:(NSString *)emailText;
+
+/** @fn didChangeEmail:
+    @brief Should be called after any change of email value. Updates UI controls state
```
[the remark this answers is not in the document]
Comment (renkelvin) [reply FirebaseEmailAuthUI/FUIConfirmEmailViewController.h]: Yeah, it's called in `textFieldDidChange`

**Comment by renkelvin** [comment]
Working on the Travis issue.

File: FirebaseEmailAuthUI/FUIConfirmEmailViewController.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Yue-Wang-Google) [reply FirebaseEmailAuthUI/FUIConfirmEmailViewController.h]: I mean you need to reverse the @brief statement.

File: FirebaseEmailAuthUI/FUIConfirmEmailViewController.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yue-Wang-Google) [reply FirebaseEmailAuthUI/FUIConfirmEmailViewController.h]: remove this line

even if you want to keep it, indent it after @brief, and use e.g.

[the remark this answers is not in the document]
Comment (renkelvin) [reply FirebaseEmailAuthUI/FUIConfirmEmailViewController.h]: OK, I was confused because it quoted 3 lines.

**Comment by renkelvin** [comment]
Thank you!

**Comment by renkelvin** [comment]
The pr became unmanageable after rebasing. Moved to a new clean pr: https://github.com/firebase/FirebaseUI-iOS/pull/588

**Pull request closed without merging**